### PR TITLE
Extract job configuration change event handler SPI and impl in PipelineJobExecutor for common usage

### LINKED ISSUE #20730: Extract job configuration change event handler SPI and impl in PipelineJobExecutor for common usage
Part of #19421.

Currently, it's hard-coded. It's better to extract job configuration change event handler SPI for common usage. SPI type could be JobType.typeName.

Related code:
- PipelineJobExecutor.processEvent method

Possible changes:
- Add SPI definition
- Add SPI factory
- Add SPI impl for migration job

### PR DESCRIPTION
**Pull request by azexcy**
Fixes #20730.

Changes proposed in this pull request:
- Add SPI and factory
-

### CODE REVIEW

File: shardingsphere-kernel/shardingsphere-data-pipeline/shardingsphere-data-pipeline-core/src/main/java/org/apache/shardingsphere/data/pipeline/core/execute/PipelineJobExecutor.java [hunk 1]
```diff
@@ -75,16 +74,16 @@ private Optional<JobConfigurationPOJO> getJobConfigPOJO(final DataChangedEvent e
...
-                new MigrationJobPreparer().cleanup(jobConfig);
-            } else if (PipelineJobProgressDetector.isJobSuccessful(jobConfig.getJobShardingCount(), MigrationJobAPIFactory.getInstance().getJobProgress(jobConfig).values())) {
+                process.cleanup(jobConfigPOJO.getJobParameter());
+            } else if (process.isJobSuccessful(jobConfigPOJO.getJobParameter())) {
                 log.info("isJobSuccessful=true");
-                new MigrationJobPreparer().cleanup(jobConfig);
+                process.cleanup(jobConfigPOJO.getJobParameter());
             }
             PipelineJobCenter.stop(jobId);
             return;
```
Comment (sandynz) [inline shardingsphere-kernel/shardingsphere-data-pipeline/shardingsphere-data-pipeline-core/src/main/java/org/apache/shardingsphere/data/pipeline/core/execute/PipelineJobExecutor.java:89]: After getting `jobType`, we could delegate all implementation to SPI and impl, so new type of job is free to implement it as needed.

File: shardingsphere-kernel/shardingsphere-data-pipeline/shardingsphere-data-pipeline-api/src/main/java/org/apache/shardingsphere/data/pipeline/spi/process/JobConfigEventProcess.java [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+    void cleanup(String jobParameter);
+    
+    /**
+     * Check job is successful.
+     *
+     * @param jobParameter job parameter
+     * @return true if job is successful otherwise is false
+     */
+    boolean isJobSuccessful(String jobParameter);
+}
```
Comment (sandynz) [inline shardingsphere-kernel/shardingsphere-data-pipeline/shardingsphere-data-pipeline-api/src/main/java/org/apache/shardingsphere/data/pipeline/spi/process/JobConfigEventProcess.java:41]: SPI name and methods could be updated, it's too detailed, could to be higher level.

File: shardingsphere-kernel/shardingsphere-data-pipeline/shardingsphere-data-pipeline-api/src/main/java/org/apache/shardingsphere/data/pipeline/spi/process/JobConfigEventProcessFactory.java [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+    static {
+        ShardingSphereServiceLoader.register(JobConfigEventProcess.class);
+    }
+    
+    /**
+     * Get importer creator instance.
+     *
+     * @param jobType job type
+     * @return importer creator
+     */
```
Comment (sandynz) [inline shardingsphere-kernel/shardingsphere-data-pipeline/shardingsphere-data-pipeline-api/src/main/java/org/apache/shardingsphere/data/pipeline/spi/process/JobConfigEventProcessFactory.java:41]: Looks javadoc doesn't match method

**Review by sandynz — CHANGES_REQUESTED** [review]

File: shardingsphere-kernel/shardingsphere-data-pipeline/shardingsphere-data-pipeline-core/src/main/java/org/apache/shardingsphere/data/pipeline/core/spi/process/PipelineEventProcess.java [hunk 4]
```diff
@@ -0,0 +1,41 @@
...
+     * @param jobConfigPOJO job config
+     */
+    void deleteEventHandle(JobConfigurationPOJO jobConfigPOJO);
+    
+    /**
+     * Disable event handle.
+     *
+     * @param jobConfigPOJO job config
+     */
+    void disableEventHandle(JobConfigurationPOJO jobConfigPOJO);
```
Comment (sandynz) [inline shardingsphere-kernel/shardingsphere-data-pipeline/shardingsphere-data-pipeline-core/src/main/java/org/apache/shardingsphere/data/pipeline/core/spi/process/PipelineEventProcess.java:40]: Could we merge these methods?

File: shardingsphere-kernel/shardingsphere-data-pipeline/shardingsphere-data-pipeline-core/src/main/java/org/apache/shardingsphere/data/pipeline/core/execute/PipelineJobExecutor.java [hunk 5]
```diff
@@ -73,20 +72,17 @@ private Optional<JobConfigurationPOJO> getJobConfigPOJO(final DataChangedEvent e
...
+        PipelineEventProcess process = PipelineEventProcessFactory.getInstance(jobType);
+        if (isDeleted) {
+            process.deleteEventHandle(jobConfigPOJO);
+            return;
+        }
+        if (isDisabled) {
+            process.disableEventHandle(jobConfigPOJO);
             return;
         }
         switch (event.getType()) {
```
Comment (sandynz) [inline shardingsphere-kernel/shardingsphere-data-pipeline/shardingsphere-data-pipeline-core/src/main/java/org/apache/shardingsphere/data/pipeline/core/execute/PipelineJobExecutor.java:88]: These code could be put to SPI impl

File: shardingsphere-kernel/shardingsphere-data-pipeline/shardingsphere-data-pipeline-core/src/main/java/org/apache/shardingsphere/data/pipeline/core/spi/process/PipelineEventProcess.java (hunk 4 above)
Comment (sandynz) [reply shardingsphere-kernel/shardingsphere-data-pipeline/shardingsphere-data-pipeline-core/src/main/java/org/apache/shardingsphere/data/pipeline/core/spi/process/PipelineEventProcess.java:40]: e.g. void process(DataChangedEvent event, JobConfigurationPOJO jobConfigPOJO);

File: shardingsphere-kernel/shardingsphere-data-pipeline/shardingsphere-data-pipeline-core/src/main/java/org/apache/shardingsphere/data/pipeline/core/spi/process/PipelineEventProcess.java [hunk 6]
```diff
@@ -0,0 +1,41 @@
...
+
+package org.apache.shardingsphere.data.pipeline.core.spi.process;
+
+import org.apache.shardingsphere.elasticjob.infra.pojo.JobConfigurationPOJO;
+import org.apache.shardingsphere.infra.util.spi.type.typed.TypedSPI;
+
+/**
+ * Job config event process.
+ */
+public interface PipelineEventProcess extends TypedSPI {
```
Comment (sandynz) [inline shardingsphere-kernel/shardingsphere-data-pipeline/shardingsphere-data-pipeline-core/src/main/java/org/apache/shardingsphere/data/pipeline/core/spi/process/PipelineEventProcess.java:26]: Could we improve class name? e.g. PipelineJobConfigurationChangedEventProcessor

File: shardingsphere-kernel/shardingsphere-data-pipeline/shardingsphere-data-pipeline-core/src/main/java/org/apache/shardingsphere/data/pipeline/core/execute/PipelineJobExecutor.java (hunk 5 above)
Comment (sandynz) [reply shardingsphere-kernel/shardingsphere-data-pipeline/shardingsphere-data-pipeline-core/src/main/java/org/apache/shardingsphere/data/pipeline/core/execute/PipelineJobExecutor.java:88]: Maybe we could finish another related refactor firstly

Comment (sandynz) [reply shardingsphere-kernel/shardingsphere-data-pipeline/shardingsphere-data-pipeline-core/src/main/java/org/apache/shardingsphere/data/pipeline/core/execute/PipelineJobExecutor.java:88]: Maybe we could finish another related refactor firstly

**Review by sandynz — CHANGES_REQUESTED** [review]

**Comment by azexcy** [comment]
Another pr will resolve the same feature. so closed.

**Review by sandynz — APPROVED** [review]

**Pull request merged**